release_util/management/commands/run_migrations.py

```diff
@@ -0,0 +1,180 @@
+            time_to_fail = self.timer() - start
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+            self.stderr.write("Migration failed for app '{}' - migration '{}'.\n".format(app, migration))
+            self.migration_state['failure'][(app, migration)] = {
+                'duration': time_to_fail,
+                'output': out.getvalue(),
+                'traceback': repr(traceback.format_exception(exc_type, exc_value, exc_traceback))
+            }
+            # Add the remaining migrations to the unapplied status.
+            while self.more_to_apply():
+                app_migration = self.to_apply.pop(0)
+                self.migration_state['unapplied'].append(app_migration)
+            # Find the apps that have been applied -or- failed.
+            # Include their initial migrations as commands.
+            apps_to_rollback = []
+            for state in ['success', 'failure']:
+                apps_to_rollback += [k[0] for k in self.migration_state[state].keys()]
+            apps_to_rollback = list(set(apps_to_rollback))
+            for app in apps_to_rollback:
+                self.migration_state['rollback_commands'].append(
+                    [
+                        'python', 'manage.py', 'migrate',
+                        app,
+                        self.input_migrations['initial_states'][app]
+                    ]
+                )
+            raise CommandError()
+
+        time_to_apply = self.timer() - start
+        self.migration_state['success'][(app, migration)] = {
+            'duration': time_to_apply,
+            'output': out.getvalue(),
+        }
+
+    def apply_all(self):
+        """
+        Apply all the migrations.
+        """
+        while self.more_to_apply():
+            self._apply_next()
+
+    def state(self):
+        """
+        Returns the state of the current
+        """
+        state = deepcopy(self.migration_state)
+        for key in ('success', 'failure'):
+            state[key] = dict(state[key])
+        return yaml.dump(state)
+
+
+class Command(BaseCommand):
+    """
+    Given a YAML file containing apps and migrations, apply those app migrations.
+    The YAML format for the apps/migrations is:
+
+    migrations:
+      - [app1, 0001_initial]
+      - [app2, 0012_otherthing]
+      - [app1, 0002_somthing]
+    initial_states:
+      - app1:
+        - zero
+      - app2:
+        - 0011_beforetheotherthing
+
+    If an error occurs in any of the migrations, the migrations are halted at that point
+        and the status is recorded in an artifact.
+    The output of the command is in YAML format and specifies the following:
+        - migrations that run successfully and how long they took
+        - migrations that failed and how long the failure took
+        - migrations that unapplied due to previous failures
+    The output YAML format:
+
+    success:
+      ? !!python/tuple [app1, 0001_initial]
```

Comment:
The reason I used tuples here was because I was using them as keys in a dictionary - and Python doesn't allow lists (mutables) to be keys.

I'll attempt to remove the tuples somehow. I agree that it fouls up the YAML output.